internal/pkg/controllers/profilerecorder/profilerecorder.go

```diff
@@ -0,0 +1,256 @@
+	reasonProfileCreated        event.Reason = "SeccompProfileCreated"
+	reasonProfileCreationFailed event.Reason = "CannotCreateSeccompProfile"
+	reasonAnnotationParsing     event.Reason = "SeccompAnnotationParsing"
+)
+
+type RecorderReconciler struct {
+	client        client.Client
+	log           logr.Logger
+	record        event.Recorder
+	nodeAddresses []string
+	podsToWatch   map[string]string
+}
+
+func Setup(ctx context.Context, mgr ctrl.Manager, l logr.Logger) error {
+	const name = "profilerecorder"
+	c, err := client.New(mgr.GetConfig(), client.Options{})
+	if err != nil {
+		return errors.Wrap(err, errGetClient)
+	}
+
+	node := &corev1.Node{}
+	if c.Get(
+		context.Background(),
+		client.ObjectKey{Name: os.Getenv(config.NodeNameEnvKey)},
+		node,
+	) != nil {
+		return errors.Wrap(err, errGetNode)
+	}
+
+	nodeAddresses := []string{}
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			l.Info("Setting up profile recorder", "Node", addr.Address)
+			nodeAddresses = append(nodeAddresses, addr.Address)
+			break
+		}
+	}
+
+	if len(nodeAddresses) == 0 {
+		return errors.New("Unable to get node's internal Address")
+	}
+
+	reconciler := RecorderReconciler{
+		client:        mgr.GetClient(),
+		log:           l,
+		nodeAddresses: nodeAddresses,
+		record:        event.NewAPIRecorder(mgr.GetEventRecorderFor(name)),
+		podsToWatch:   map[string]string{},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithEventFilter(predicate.And(
+			resource.NewPredicates(reconciler.isPodWithTraceAnnotation),
+			resource.NewPredicates(reconciler.isPodOnLocalNode),
+		)).
+		For(&corev1.Pod{}).
+		Complete(&reconciler)
+}
+
+func (r *RecorderReconciler) isPodOnLocalNode(obj runtime.Object) bool {
+	p, ok := obj.(*corev1.Pod)
+
+	if !ok {
+		return false
+	}
+
+	for _, addr := range r.nodeAddresses {
+		if p.Status.HostIP == addr {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (r *RecorderReconciler) isPodWithTraceAnnotation(obj runtime.Object) bool {
+	p, ok := obj.(*corev1.Pod)
+
+	if !ok {
+		return false
+	}
+
+	return p.Annotations[config.SeccompProfileRecordAnnotationKey] != ""
+}
+
+func (r *RecorderReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	logger := r.log.WithValues("recorder", req.Name, "namespace", req.Namespace)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	pod := &corev1.Pod{}
+	if err := r.client.Get(ctx, req.NamespacedName, pod); err != nil {
+		if kerrors.IsNotFound(err) {
+			if err := r.collectProfile(ctx, req.NamespacedName); err != nil {
+				return reconcile.Result{}, nil
+			}
+		} else {
+			// Returning an error means we will be requeued implicitly.
+			logger.Error(err, "Error reading pod")
+			return reconcile.Result{}, errors.Wrap(err, errGetPod)
+		}
+	}
+
+	if pod.Status.Phase == corev1.PodRunning || pod.Status.Phase == corev1.PodSucceeded {
```

Comment:
if the pod already succeeded why would we start recording at this point? Wouldn't we already loose all the container's events then?

Reply:
Ah yeah I think we can remove that code path, too.